### ISSUE TITLE
AMQP_Protocol Connection Does Not Reconnect

### ISSUE THREAD

**Issue by mv-bp** [Azure/azure-iot-sdk-c#628]
- **OS and version used:**
Operating System:  Running on a linux variation using Kernel 3.12.70
Processor:		32bit ARM5 

- **SDK version used:** 
azure-iot-sdk-c 1.2.4

# Description of the issue:
This is happening consistently across devices running our application using the Azure IoT C SDK using the AMQP_Protocol.  
The application on each device connects, sends a message every 5 minutes, and asynchronously receive messages from the Azure IoT hub.
After 49 days and 17 hours (consistent across all of our devices) the connection appears to fail.  
After the 49 days and 17 hours, the send callback result is IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT.  
Subsequently, the application continues to attempt to send a message every 5 minutes and the send callback is a MESSAGE TIMEOUT.
The connection callback never returns a NO_NETWORK or UNAUTHENTICATED which could indicate that the SDK does not know that the connection is no longer valid and thus does not attempt to reconnect.
Only after an application restart is the connection re-established for another 49 days and 17 hours.

Checking the application logs, the application continues to operate (no seg faults, no application crashes, no adverse memory and no adverse CPU usage) attempting to send messages every 5 minutes.

This issue is easy to reproduce however takes 49 days and 17 hours.

The SDK log trace:
Unable to capture trace due to length of time to reproduce

Connection Protocol: 		AMQP_Protocol
Connection Retry Policy: 	IOTHUB_CLIENT_RETRY_INTERVAL

Message Timeout:			5000 (milliseconds = 5 seconds)
SasTokenLifetime:			86400 (seconds = 1 day = 24 hours = 24 hours/day * 60 minutes/hour * 60 seconds/minute = max)

Environment:
Azure IoT SDK C version: 	azure-iot-sdk-c 1.2.4
Operating System:  		Running on a linux variation using Kernel 3.12.70
Processor:				32bit ARM5 

# Console log of the issue:
Unable to capture trace due to length of time to reproduce

**Comment by mv-bp** [comment]
Potentially is this the issue in the SDK??

typedef uint_fast32_t tickcounter_ms_t;
typedef unsigned int       uint_fast32_t;
if (tickcounter_get_current_ms(connection->tick_counter, &current_ms) != 0)

32 bit MAX
0x FFFFFFFF = 4294967295
4294967295 milliseconds = 4294967 seconds = 1193 hours = 49 days (approx.)

Thus causing an overflow to happen which messes up the connection logic time related deadlines throwing things out of gear??

https://github.com/Azure/azure-uamqp-c/blob/c7770c1a45a62d451353398e281fb5a661cc655c/src/connection.c#L1755

**Comment by jebrando** [comment]
@mv-bp the fix for this issue has been checked in, please either reopen this issue if you find any other issue witht he fix, or open an new issue.

Thanks

**Comment by az-iot-builder-01** [comment]
@mv-bp, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
